File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/roadrunner/PinpointLocalizer.java
package org.firstinspires.ftc.teamcode.roadrunner;

import androidx.annotation.NonNull;

import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.roadrunner.Pose2d;
import com.acmerobotics.roadrunner.PoseVelocity2d;
import com.acmerobotics.roadrunner.Rotation2d;
import com.acmerobotics.roadrunner.Vector2d;
import com.acmerobotics.roadrunner.ftc.PinpointView;
import com.qualcomm.robotcore.hardware.DcMotorSimple;
import com.qualcomm.robotcore.hardware.HardwareMap;

import java.util.Objects;

/** @noinspection unused*/
@Config
public final class PinpointLocalizer implements Localizer {
    public static class Params {
        public double parYTicks = 0.0; // y position of the parallel encoder (in tick units)
        public double perpXTicks = 0.0; // x position of the perpendicular encoder (in tick units)
    }

    public static Params PARAMS = new Params();
    public final PinpointView view;

    private final GoBildaPinpointDriver driver;

    private Pose2d txWorldPinpoint;
    private Pose2d txPinpointRobot = new Pose2d(0, 0, 0);

    public PinpointLocalizer(HardwareMap hardwareMap, Pose2d initialPose) {
        // TODO: make sure your config has a Pinpoint device with this name
        //   see https://ftc-docs.firstinspires.org/en/latest/hardware_and_software_configuration/configuring/index.html
        driver = hardwareMap.get(GoBildaPinpointDriver.class, "pinpoint");

        //double mmPerTick = 25.4 * inPerTick;
        //driver.setEncoderResolution(1 / mmPerTick);
        //driver.setOffsets(mmPerTick * PARAMS.parYTicks, mmPerTick * PARAMS.perpXTicks);
        //NOTE FROM DRAGOS: we already know the encoder resolution for the gobilda 4 bar pod
        //  so we're gonna skip odo tuning completely, setting the inPerTick from MecanumDrive
        //  to inches per millimeter, and ignoring it here
        double mmPerInch = 25.4;
        driver.setEncoderResolution(GoBildaPinpointDriver.GoBildaOdometryPods.goBILDA_4_BAR_POD);
        driver.setOffsets(mmPerInch * PARAMS.parYTicks, mmPerInch * PARAMS.perpXTicks);

        // TODO: reverse encoder directions if needed
        //    driver.setEncoderDirections(GoBildaPinpointDriver.EncoderDirection.REVERSED, GoBildaPinpointDriver.EncoderDirection.REVERSED);

        driver.resetPosAndIMU();

        txWorldPinpoint = initialPose;

        view = getView();
    }

    @Override
    public void setPose(Pose2d pose) {
        txWorldPinpoint = pose.times(txPinpointRobot.inverse());
    }

    @Override
    public Pose2d getPose() {
        return txWorldPinpoint.times(txPinpointRobot);
    }

    @Override
    public PoseVelocity2d update() {
        driver.update();
        if (Objects.requireNonNull(driver.getDeviceStatus()) == GoBildaPinpointDriver.DeviceStatus.READY) {
            txPinpointRobot = new Pose2d(driver.getPosX() / 25.4, driver.getPosY() / 25.4, driver.getHeading());
            Vector2d worldVelocity = new Vector2d(driver.getVelX() / 25.4, driver.getVelY() / 25.4);
            Vector2d robotVelocity = Rotation2d.fromDouble(-driver.getHeading()).times(worldVelocity);
            return new PoseVelocity2d(robotVelocity, driver.getHeadingVelocity());
        }
        return new PoseVelocity2d(new Vector2d(0, 0), 0);
    }

    private PinpointView getView() {
        return new PinpointView() {
            @Override
            public void update() {
                driver.update();
            }

            @Override
            public int getParEncoderPosition() {
                return driver.getEncoderX();
            }

            @Override
            public int getPerpEncoderPosition() {
                return driver.getEncoderY();
            }

            @Override
            public float getHeadingVelocity() {
                return (float) driver.getHeadingVelocity();
            }

            //NOTE: not used by tuning
            @Override
            public void setParDirection(@NonNull DcMotorSimple.Direction direction) {}

            @Override
            public void setPerpDirection(@NonNull DcMotorSimple.Direction direction) {}
        };
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/roadrunner/PinpointLocalizer.kt
package org.firstinspires.ftc.teamcode.roadrunner

import com.acmerobotics.dashboard.config.Config
import com.acmerobotics.roadrunner.Pose2d
import com.acmerobotics.roadrunner.PoseVelocity2d
import com.acmerobotics.roadrunner.Vector2d
import com.acmerobotics.roadrunner.ftc.RawEncoder
import com.qualcomm.hardware.lynx.LynxDcMotorController
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.configuration.LynxConstants
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit
import org.firstinspires.ftc.robotcore.external.navigation.DistanceUnit
import org.firstinspires.ftc.robotcore.external.navigation.Pose2D
import org.firstinspires.ftc.teamcode.library.pinpoint.GoBildaPinpointDriver
import org.firstinspires.ftc.teamcode.library.pinpoint.fakeOdoMotor

class PinpointLocalizer @JvmOverloads constructor(
    hardwareMap: HardwareMap,
    initPose: Pose2d = Pose2d(0.0, 0.0, 0.0),
    val telemetry: Telemetry? = null
): Localizer {
    @Config
    data object PinpointConfig {
        @JvmField
        var xOffset = 23.0
        @JvmField
        var yOffset = -115.0
    }

    private var currentPose: Pose2d = initPose
    val odo: GoBildaPinpointDriver = hardwareMap.get(GoBildaPinpointDriver::class.java, "odo")

    val encX: RawEncoder
    val encY: RawEncoder

    init {
        val motorController = hardwareMap.getAll(LynxDcMotorController::class.java).first {
            LynxConstants.isEmbeddedSerialNumber(it.serialNumber)
        }
        encX = RawEncoder(fakeOdoMotor(
            { odo.posX },
            { odo.velX },
            { odo.update() },
            motorController
        ))
        encY = RawEncoder(fakeOdoMotor(
            { odo.posY },
            { odo.velY },
            { odo.update() },
            motorController
        ))

        odo.setOffsets(
            PinpointConfig.xOffset,
            PinpointConfig.yOffset
        )
        odo.setEncoderResolution(GoBildaPinpointDriver.GoBildaOdometryPods.goBILDA_4_BAR_POD)
        odo.setEncoderDirections(
            GoBildaPinpointDriver.EncoderDirection.FORWARD,
            GoBildaPinpointDriver.EncoderDirection.REVERSED
        )

        /*val initLine = telemetry?.addLine("Reseting pinpoint")
        telemetry?.update()
        odo.resetPosAndIMU()

        while (odo.deviceStatus != GoBildaPinpointDriver.DeviceStatus.READY) {
            try {
                Thread.sleep(10)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }
        }
        telemetry?.removeLine(initLine)
        telemetry?.addLine("Pinpoint initialized")
        telemetry?.update()*/

        odo.resetPosAndIMU()

        try {
            Thread.sleep(300)
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
        }

        odo.setPosition(Pose2D(DistanceUnit.INCH, initPose.position.x, initPose.position.y, AngleUnit.RADIANS, initPose.heading.toDouble()))
    }

    override fun setPose(pose: Pose2d) {
        currentPose = pose
        odo.setPosition(Pose2D(DistanceUnit.INCH, pose.position.x, pose.position.y, AngleUnit.RADIANS, pose.heading.toDouble()))
    }

    override fun getPose(): Pose2d = currentPose

    override fun update(): PoseVelocity2d {
        odo.update()

        val odoPos = odo.position
        val odoPosX = odoPos.getX(DistanceUnit.INCH)
        val odoPosY = odoPos.getY(DistanceUnit.INCH)
        val odoPosHeading = odoPos.getHeading(AngleUnit.RADIANS)
        val odoVel = odo.velocity
        val odoVelX = odoVel.getX(DistanceUnit.INCH)
        val odoVelY = odoVel.getY(DistanceUnit.INCH)
        val odoVelHeading = odoVel.getHeading(AngleUnit.RADIANS)

        currentPose = Pose2d(
            Vector2d(
                odoPosX,
                odoPosY
            ),
            odoPosHeading
        )

        return PoseVelocity2d(
            Vector2d(
                odoVelX,
                odoVelY
            ),
            odoVelHeading
        )
    }
}